File: 프로그래머스/level3/표 편집.cpp
#include <string>
#include <stack>
#include <vector>
#include <iostream>

using namespace std;

struct Node {
    int value;
    int up;
    int down;
    bool live;
};

Node node[1000001];
stack<int> deleted;

int Up(int start, int cnt){
    int now = start;
    while(cnt>0){
        cnt--;
        now = node[now].up;
    }
    return now;
}

int Down(int start, int cnt){
    int now = start;
    while(cnt>0){
        cnt--;
        now = node[now].down;
    }
    return now;
}

int Delete(int target){
    int upNode = node[target].up;
    int downNode = node[target].down;
    node[target].live = false;
    deleted.push(target);
    if(upNode!=-1)
        node[upNode].down = downNode;
    if(downNode!=-1){
        node[downNode].up = upNode;
        return downNode;
    }
    return upNode;
}

void Recovery(){
    int target = deleted.top();
    deleted.pop();
    node[target].live = true;
    int upNode = node[target].up;
    int downNode = node[target].down;
    if(upNode != -1)
        node[upNode].down = target;
    if(downNode != -1)
        node[downNode].up = target;
}

string solution(int n, int k, vector<string> cmd) {
    string answer = "";
    int now = k;
    for(int i=0; i<n; i++){
        node[i].value = i;
        node[i].up = i-1;
        node[i].down = i+1;
        node[i].live = true;
    }
    node[n-1].down = -1;
    for(int i=0; i<cmd.size(); i++){
        if(cmd[i][0] == 'U'){
            int cnt = stoi(cmd[i].substr(2));
            now = Up(now, cnt);
        }
        else if(cmd[i][0] == 'D'){
            int cnt = stoi(cmd[i].substr(2));
            now = Down(now, cnt);
        }
        else if(cmd[i][0] == 'C'){
            now = Delete(now);
        }
        else{
            Recovery();
        }
    }
    for(int i=0; i<n; i++){
        if(node[i].live)    answer += "O";
        else                answer += "X";
    }
    return answer;
}

File: 프로그래머스/level3/표편집.js
//구현
let node = [];
let deleted = [];
let N;

const Move = (type, now, cnt) => {
    let next = now;
    if(type === 'U'){
        for(let i=0; i<cnt; i++){
            next = node[next].up;
        }
    }
    else{
        for(let i=0; i<cnt; i++){
            next = node[next].down;
        }
    }
    return next;
}

const Delete = (now) => {
    node[now].live = false;
    deleted.push(now);
    let up = node[now].up;
    let down = node[now].down;
    if(up !== -1){
        node[up].down = down;
    }
    if(down !== -1){
        node[down].up = up;
        return down;
    }
    return up;
}

const Recovery = () => {
    let last = deleted[deleted.length - 1];
    deleted.pop();
    node[last].live = true;
    let up = node[last].up;
    let down = node[last].down;
    if(up !== -1){
        node[up].down = last;
    }
    if(down !== -1){
        node[down].up = last;
    }
}

function solution(n, k, cmd) {
    var answer = "";
    N = n;
    let now = k;
    for(let i=0; i<N; i++){
        let info = {
            id: i,
            up: i-1,
            down: i+1,
            live: true
        };
        node.push(info);
    }
    node[N-1].down = -1;
    cmd.forEach(str => {
        let info = str.split(' ');
        if(info[0] === 'U' || info[0] === 'D'){
            let cnt = Number(info[1]);
            now = Move(info[0], now, cnt);
        }
        else if(info[0] === 'C'){
            now = Delete(now);
        }
        else{
            Recovery();
        }
    });
    for(let i=0; i<N; i++){
        if(node[i].live){
            answer += "O";
        }
        else{
            answer += "X";
        }
    }
    return answer;
}